promql/promql_test.go

```diff
@@ -34,3 +40,60 @@ func TestEvaluations(t *testing.T) {
 		})
 	}
 }
+
+// Run a lot of queries at the same time, to check for race conditions.
+func TestConcurrentRangeQueries(t *testing.T) {
+	stor := teststorage.New(t)
+	defer stor.Close()
+	opts := EngineOpts{
+		Logger:     nil,
+		Reg:        nil,
+		MaxSamples: 50000000,
+		Timeout:    100 * time.Second,
+	}
+	engine := NewEngine(opts)
+
+	const interval = 10000 // 10s interval.
+	// A day of data plus 10k steps.
+	numIntervals := 8640 + 10000
+	err := setupRangeQueryTestData(stor, engine, interval, numIntervals)
+	require.NoError(t, err)
+
+	cases := rangeQueryCases()
+
+	// Limit the number of queries running at the same time.
+	const numConcurrent = 4
+	sem := make(chan struct{}, numConcurrent)
+	for i := 0; i < numConcurrent; i++ {
+		sem <- struct{}{}
+	}
+	var g errgroup.Group
+	for _, c := range cases {
+		c := c
+		if strings.Contains(c.expr, "count_values") && c.steps > 10 {
+			continue // This test is too big to run with -race.
+		}
+		g.Go(func() error {
+			<-sem
+			defer func() {
+				sem <- struct{}{}
+			}()
+			qry, err := engine.NewRangeQuery(
```

Comment:
Is this call part of the race detection? Else I would prefer it to be done before the `<-sem`.

Reply:
That would have the effect that we create all queries in parallel, and hold them all in memory, then execute them over time with max 4 running in parallel.
Can you say a bit more why you would prefer that?

Reply:
Now I think about it, I might put the `<-sem` before `g.Go`, so goroutines are not created until there is a slot for them to run.

Reply:
> Can you say a bit more why you would prefer that?

NewRangeQuery will have some overhead (parsing the query etc). If we are strictly trying to find any race between the Exec calls, the overhead coming in between might not be desired, since they can prevent Exec calls from overlapping for a longer time. By putting it behind the <-sem, we could be done with the overhead before we are ready to execute.

Reply:
> I might put the <-sem before g.Go, so goroutines are not created until there is a slot for them to run.

I don't know how long the queries run, but I expect it is not a lot? In which case, IMO it is better to keep the goroutine ready.

Reply:
Then LGTM. Unless you want to get <-sem out of that goroutine, we can merge this.